[color-swatch][color-scale] First stab at implementing the coords prop

DmitrySharabin opened the pull request:
It's a very basic implementation, so we could have something to start with. 🙂
It looks like this:

<img width="1301" alt="SCR-20240607-bwjc" src="https://github.com/color-js/elements/assets/9166277/10ab30c4-73bf-4e00-8934-64ae913444dd">

When defining the type of `coords`, I used a workaround because of [the issue](https://github.com/nudeui/element/issues/27) I faced.

I wonder whether we should do something like `+(value).toPrecision(4)` to make the output look nicer (or is it OK to see that many digits?). Is there a way to specify precision in `Color.js` we could use?

**UPDATE**

Here is the code that produces the image above:


```html
<color-swatch coords="oklch.l, oklch.c, oklch.h" size="large">oklch(80% 50% 70)</color-swatch>

<div id="future_swatch_container"></div>
<script>
	let swatch = document.createElement("color-swatch");
	swatch.color = "oklch(65% 0.15 210)";
	swatch.setAttribute("size", "large");
	swatch.textContent = "Turquoise";
	swatch.coords = "L: oklch.l, C: oklch.c, H: oklch.h";
	future_swatch_container.append(swatch);
</script>

<color-scale colors="#e3fafc, #0b7285" steps="4" space="oklch" coords="L: oklch.l, C: oklch.c"></color-scale>

<color-scale space="oklch" colors="
	Gray 50: #f9fafb,
	Gray 100: #f3f4f6,
	Gray 200: #e5e7eb,
	Gray 300: #d1d5db,
	Gray 400: #9ca3af
" coords="srgb.r"></color-scale>
```

LeaVerou commented on src/color-scale/color-scale.js:
```diff
@@ -144,7 +147,11 @@ const Self = class ColorScale extends NudeElement {
 				return colors;
-			}
+			},
+			dependencies: ["colors", "steps", "space", "coords"],
```
We should have an `additionalDependencies` property rather than having to redefine all of them. Can you file it?

DmitrySharabin replied:
Done!

https://github.com/nudeui/element/issues/28

LeaVerou commented on src/color-scale/color-scale.js:
```diff
@@ -144,7 +147,11 @@ const Self = class ColorScale extends NudeElement {
+			dependencies: ["colors", "steps", "space", "coords"],
+		},
+		coords: {
+			type: String,
```
Why is this `type: String`? At the very least it should be an array or object.

LeaVerou replied:
Oh I see, because it just passes itself verbatim to `<color-swatch>`. You don't need `type: String` then, just don't include a type at all. In the future we'll probably need to parse it in `<color-scale>` though, since it could be useful for `<color-scale>` to have summary stats on these coords. But for now just passing it through is ok.

LeaVerou commented on src/color-swatch/color-swatch.js:
```diff
@@ -109,6 +111,26 @@ const Self = class ColorSwatch extends NudeElement {
 
+		if (name === "coords") {
+			if (this.color) {
+				this._el.coords?.replaceChildren(); // remove all children
```
Nit: You don't need `replaceChildren()` to remove all children, `el.textContent = ""` works and is more common.
Why the optional chaining? Is there any chance this runs before `this._el.coords` is defined? I thought we fixed that?

DmitrySharabin replied:
> Nit: You don't need `replaceChildren()` to remove all children, `el.textContent = ""` works and is more common.

This was my first guess. But I wasn't sure it was a common practice. Now I know. Thanks!

> Why the optional chaining? Is there any chance this runs before `this._el.coords` is defined? I thought we fixed that?

It's. My bad. Will remove it.

LeaVerou commented on src/color-swatch/color-swatch.js:
```diff
@@ -109,6 +111,26 @@ const Self = class ColorSwatch extends NudeElement {
+					let [space, channel] = properties.split(".");
+					let value;
+
+					if (!channel) {
+						value = this.color[space];
+					}
+					else {
+						value = this.color[space][channel];
+					}
```
```suggestion
					let value = this.color[space];

					if (channel) {
						value = value[channel];
					}
```

LeaVerou commented on src/color-swatch/color-swatch.js:
```diff
@@ -109,6 +111,26 @@ const Self = class ColorSwatch extends NudeElement {
+						value = this.color[space][channel];
+					}
+
+					this._el.coords.insertAdjacentHTML("beforeend", `<div class="coord"><dt>${ label }</dt><dd>${ value }</dd></div>`);
```
Let's minimize DOM operations by building this string and then setting the element’s entire `innerHTML` to it. You can build an array with each entry and then set innerHTML to that array with `.join("\n")`

DmitrySharabin replied:
More than agreed. 🙂

LeaVerou commented on src/color-swatch/color-swatch.js:
```diff
@@ -143,6 +165,28 @@ const Self = class ColorSwatch extends NudeElement {
+							space = this.color?.space.id;
+						}
+
+						return Color.Space.get(space)?.coords[channel]?.name ?? channel;
```
Please use `Color.Space.resolveCoord()` which exists exactly for this purpose: https://github.com/color-js/color.js/blob/main/src/ColorSpace.js#L370

DmitrySharabin replied:
With great pleasure! TIL. Thanks for sharing it with me.

LeaVerou commented on src/color-swatch/color-swatch.js:
```diff
@@ -143,6 +165,28 @@ const Self = class ColorSwatch extends NudeElement {
+			reflect: {
+				from: true,
+			},
+			dependencies: ["color"],
```
So you found a bug in Nude Element: that `defaultKey()` (and `defaultValue()`) isn't taken into account when calculating dependencies. Don't just work around bugs, file them so we can fix them!

DmitrySharabin replied:
You think about me too well! 😂 I didn't work around the bug; I knew we infer dependencies from `default`, `get`, and `convert`. And for me, it's just like that. So, I just worked with the current implementation. I'm happy that thanks to my “discovery,” we'll make `NudeElement` even better, though. 🙂

LeaVerou replied:
ANY tool you use can have bugs. That includes popular libraries, frameworks, even the browser itself. Always try to report them. And here we can even fix them at the source!

DmitrySharabin commented on src/color-swatch/color-swatch.css:
```diff
@@ -49,6 +49,33 @@ slot {
+#coords {
+	margin: 0;
+	display: inline-flex;
+	gap: .5em;
```
Agreed!

LeaVerou replied:
Actually, how hard would it be to not have the `<dl>` at all in that case?

DmitrySharabin replied:
That's a much better solution—not to have `<dl>` if unnecessary. Thank you for the idea. It should work that way from now on.

DmitrySharabin commented:
I addressed all your feedback. Thank you so much! It should be ready now for another iteration.

LeaVerou commented on src/color-swatch/color-swatch.js:
```diff
@@ -109,6 +111,27 @@ const Self = class ColorSwatch extends NudeElement {
+				let coords = [];
+				for (let coord of this.coords) {
+					let [label, properties] = Object.entries(coord)[0];
+					let [space, channel] = properties.split(".");
```
This is still parsing coords manually

DmitrySharabin replied:
Oops. Fixed.

LeaVerou reviewed (changes requested):
See comments

DmitrySharabin commented:
I addressed all your feedback. Thank you!

Before requesting your review again, I'd love to know if there is a way to output formatted values for coordinates (the way they are displayed in the parsed color). We might get a result much closer to the one you showcased in the issue:

<img width="210" alt="image" src="https://github.com/color-js/elements/assets/175836/39daa1fd-ed00-4057-acea-997a0cc84a59">

LeaVerou commented:
No, but that's an excellent feature request for `getAll()`.

LeaVerou commented on src/color-swatch/color-swatch.js:
```diff
@@ -109,6 +109,32 @@ const Self = class ColorSwatch extends NudeElement {
+			}
+
+			if (!this._el.coords) {
+				this._el.colorWrapper.insertAdjacentHTML("afterend", `<dl part="coords"></dl>`);
```
Nit but I'd rather only insert this once and then keep it around in `this._el.coords`:

```js
this._el.coords ??= Object.assign(document.createElement("dl"), {part: "coords"});
```

DmitrySharabin replied:
Q: Do we want to place `this._el.coords` inside `this._el.colorWrapper`? Previously, it was its sibling.

LeaVerou replied:
No, use `.before()` instead

DmitrySharabin commented:
> No, but that's an excellent feature request for `getAll()`.

Do you mean `Color.getAll()` in `Color.js`? Should I file an issue?

LeaVerou commented:
> > No, but that's an excellent feature request for `getAll()`.
> 
> Do you mean `Color.getAll()` in `Color.js`? Should I file an issue?

Yes

LeaVerou commented:
Don't worry about formatting for now, just apply precision. We'll deal with formatting in Color.js.

DmitrySharabin commented:
> Don't worry about formatting for now, just apply precision. We'll deal with formatting in Color.js.

Sure. Just to ensure I follow: we can do it like this: `value.toPrecision(4)` Right? Or do we have a way to get the needed precision with `Color.js` (with some feature I'm not yet aware of 🫣)?

LeaVerou commented:
Yup, `+value.toPrecision()`.
